File: basics/basics-typescript/src/6_events_state.ts
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate Examples for the Node.js/TypeScript SDK,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/blob/main/LICENSE
 */

import * as restate from "@restatedev/restate-sdk";

//
// Populate state from events (from Kafka).
// Query the state via simple RPC/HTTP calls.
//

const profileService = restate.object({
  name: "profile",
  handlers: {
    registration: async (
      ctx: restate.ObjectContext,
      event: { name: string }
    ) => {
      // store in state the user's information as coming from the registration event
      ctx.set("name", event.name);
    },

    email: async (ctx: restate.ObjectContext, event: { email: string }) => {
      // store in state the user's information as coming from the email event
      ctx.set("email", event.email);
    },

    get: async (ctx: restate.ObjectContext): Promise<UserProfile> => {
      return {
        id: ctx.key,
        name: (await ctx.get<string>("name")) ?? "",
        email: (await ctx.get<string>("email")) ?? "",
      };
    },
  },
});

type UserProfile = {
  id: string;
  name: string;
  email: string;
};

restate.endpoint().bind(profileService).listen();

// Update and query the state via:
/*
curl localhost:8080/profile/userid1/registration --json '{ "name": "Bob" }'
curl localhost:8080/profile/userid1/email --json '{ "email": "bob@mydomain.com"}'
curl localhost:8080/profile/userid1/get
*/

File: python/basics/app/6_events_state.py
import restate
from restate import VirtualObject, ObjectContext

# Populate state from events (from Kafka).
# Query the state via simple RPC/HTTP calls.
profile_service = VirtualObject("profile")


@profile_service.handler()
async def registration(ctx: ObjectContext, name: str):
    print(name)
    # store in state the user's information as coming from the registration event
    ctx.set("name", name)


@profile_service.handler()
async def email(ctx: ObjectContext, email: str):
    # store in state the user's information as coming from the email event
    ctx.set("email", email)


@profile_service.handler()
async def get(ctx: ObjectContext) -> dict:
    return {
        "id": ctx.key(),
        "name": await ctx.get("name"),
        "email": await ctx.get("email")
    }


app = restate.app(services=[profile_service])


# Start up Kafka:
# docker-compose up
#
# Run Restate with the Kafka configuration:
# restate-server --config-file restate.toml
#
# Let Restate subscribe to the users topic:
# curl localhost:9070/subscriptions -H 'content-type: application/json' \
#     -d '{
#             "source": "kafka://my-cluster/profiles",
#             "sink": "service://profile/registration",
#             "options": {"auto.offset.reset": "earliest"}
#         }'
#
# You can start a Kafka producer by exec'ing into the Kafka broker:
# docker exec -ti f7381464eaef /bin/bash
#
# kafka-console-producer --bootstrap-server localhost:9092 --topic profiles --property "parse.key=true" --property "key.separator=:"
#
# Publish the following event to Kafka:
# userid1:"Bob"
#
# Update and query the state via:
# curl localhost:8080/profile/userid1/email --json '"bob@mydomain.com"'
# curl localhost:8080/profile/userid1/get
